lib/persistent-knex-model-mixin.js

```diff
@@ -0,0 +1,123 @@
+'use strict'
+const _ = require('lodash')
+
+module.exports = function PersistentKnexModelMixin (Model, knex, inOptions, userOptions) {
+  if (typeof Model.convertFromPersistent !== 'function') {
+    Model.convertFromPersistent = function (data) {
+      return data
+    }
+  }
+
+  if (typeof Model.convertToPersistent !== 'function') {
+    Model.convertToPersistent = function (data) {
+      return data
+    }
+  }
+
+  /**
+   * Set data from database.
+   *
+   * Applies filters for conversion from database format to normal format.
+   */
+  Model.prototype.setDataPersistent = function (data) {
+    this.setData(this._applyFilter(this.constructor.convertFromPersistent, data))
+  }
+
+  Model.prototype.getDataPersistent = function () {
+    return this._applyFilter(this.constructor.convertToPersistent, this.getData())
+  }
+
+  Model.prototype.save = function * (options) {
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.update(this.getDataPersistent()).into(Model.tableName).where('id', this.id).then()
+  }
+
+  // Delete row based on where options
+  Model.prototype.destroy = function (selectOptions) {
+    return knex(Model.tableName).where(selectOptions).del()
+  }
+
+  Model.fromDatabaseModel = function (data) {
+    if (!data) return null
+    // If result is an array (return from knex select query), take the first object
+    if (Array.isArray(data)) {
+      data = data[0]
+    }
+    const model = new Model()
+    model.setDataPersistent(data)
+    return model
+  }
+
+  Model.fromDataPersistent = function (data) {
+    const model = new Model()
+    model.setDataPersistent(data)
+    return model
+  }
+
+  Model.findAll = function * findAll () {
+    // this returns an array of JSON objects
+    return knex(Model.tableName).select().then()
+  }
+  Model.findById = function * findById (id, options) {
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.from(Model.tableName).select().where('id', id).then(fromDatabaseModel)
+  }
+
+  // Optional: transaction
+  // set this to knex transaction if you want this operation to be done inside transaction.
+  Model.create = function * create (values, options) {
+    const model = Model.fromData(values)
+    const modelPersistent = model.getDataPersistent()
+    const myKnex = knexOrTransaction(knex, options)
+    return myKnex.insert(cleanUndefinedAttributes(modelPersistent)).into(Model.tableName)
+  }
+
+  Model.bulkCreate = function * bulkCreate (records, options) {
+    const models = records
+      .map(Model.fromData.bind(Model))
+      .map(_.method('getDataPersistent'))
```

Comment:
I could not make it work with the previous line's pattern.  I opened a conversation with Stefan and you about this line.  Let's move on with this code, and continue with the conversation in the background.